usr/sbin/p11sak/p11sak.c

```diff
@@ -1141,6 +1191,200 @@ static CK_ATTRIBUTE_TYPE col2type(int col)
+                    // build template
+                    CK_BBOOL a_bool;
+                    CK_ULONG bs = sizeof(CK_BBOOL);
+                    CK_ATTRIBUTE temp[] = { {hex, &a_bool, bs} };
+
+                    // get attribute value
+                    rc = funcs->C_GetAttributeValue(session, hkey, temp, 1);
+                    if (rc == CKR_OK) {
+                        if (temp[0].ulValueLen == 0) {
+                            printf("          %s: not exportable\n", confignode_to_bareval(name)->value);
+                        } else if (temp[0].ulValueLen != sizeof(CK_BBOOL)) {
+                            printf(" Error in retrieving Attribute %s: %lu\n",
+                                    confignode_to_bareval(name)->value, temp[0].ulValueLen);
+                        } else {
+                            printf("          %s: %s\n", confignode_to_bareval(name)->value,
+                                    CK_BBOOL2a(*(CK_BBOOL*) temp[0].pValue));
+                        }    
+                    } else if (rc == CKR_ATTRIBUTE_SENSITIVE) {
+                        printf("          %s: SENSITIVE\n", confignode_to_bareval(name)->value);
+                    } else if (rc != CKR_ATTRIBUTE_TYPE_INVALID) {
+                        return rc;
+                    } 
+                } else if (strcmp((confignode_to_bareval(type)->value), "CK_ULONG") == 0) {
+                    // build template
+                    CK_ULONG a_ulong;
+                    CK_ULONG us = sizeof(CK_ULONG);
+                    CK_ATTRIBUTE temp[] = { {hex, &a_ulong, us} };
+
+                    // get attribute value
+                    rc = funcs->C_GetAttributeValue(session, hkey, temp, 1);
+                    if (rc == CKR_OK) {
+                        if (temp[0].ulValueLen != sizeof(CK_ULONG)) {
+                            printf(" Error in retrieving Attribute %s: %lu\n",
+                                    confignode_to_bareval(name)->value, temp[0].ulValueLen);
+                        } else {
+                            printf("          %s: %lu (0x%lX)\n", confignode_to_bareval(name)->value, 
+                                *(CK_ULONG*) temp[0].pValue, *(CK_ULONG*) temp[0].pValue);
+                        }    
+                    } else if (rc == CKR_ATTRIBUTE_SENSITIVE) {
+                        printf("          %s: SENSITIVE\n", confignode_to_bareval(name)->value);
+                    } else if (rc != CKR_ATTRIBUTE_TYPE_INVALID) {
+                        return rc;
+                    } 
+                } else if (strcmp((confignode_to_bareval(type)->value), "CK_BYTE") == 0) {
+                    // get length 
+                    CK_ATTRIBUTE temp[] = { {hex, NULL, 0} };
+                    rc = funcs->C_GetAttributeValue(session, hkey, temp, 1);
+
+                    // build template
+                    CK_BYTE a_byte_array[temp[0].ulValueLen];
```

Comment:
Use of dynamic array is not a good idea.  In general, do a sanity check of `temp[0].ulValueLen` which might be `CK_UNAVAILABLE_INFORMATION` and thus very huge, then `malloc` the value and `free` it before it leaves scope.

Reply:
Done